### DISCUSSION TITLE
Add 'npm install --compat <pkg>' command for auto-install a version of a <pkg> that is compatible with all other versions of packages in package.json

### DISCUSSION THREAD

**Discussion by TetianaNazarova** [npm/rfcs#515] [General]
Could you, **please, consider** the possibility of adding a **'npm install --compat <pkg> package'**  or 'npm compat <pkg> package' command **for automatic installation** in package.json **a package <pkg> with a version which is compatible with all other packages versions in package.json?**

In my practice I had a lot of events when I had to add a new package to package.json (or update - increase or decrease the version of the already installed package) and then I got a **lot of errors which were caused by the incompatibility of versions of this package with other packages in package.json**. 

And I had to spent a lot of time (from several minutes to several hours or even days) for struggling with that problem - I had to select a needed version of the package <pkg> manually, iterating over versions pointed at https://www.npmjs.com/ and guided by intuition, luck and approximately the same release time of this package and other main packages in package.json. And if there is not 1 but several incompatibilities with several other packages it becomes a real pain.

If there is impossible to install a package with the existing versions of other packages in package.json maybe **there should be an output with the instructions which packages should have to change their versions and what exactly versions should be installed for each of them and what version of new added (or with changed version) package will be compatible** with them all. 

If the cause of this problem is an old / new npm or node.js version - maybe there should be an output with the instructions what version of npm and / or node.js should be installed for a proper work.

Is there a possibility to automate this process in a some way? Or it isn't possible at all?

I know, that many developers had faced with a such problem - there are a lot of GitHub issues and StackOverflow questions connected to this problem in a great variations relating to a great number of different packages (libraries). 

I guess that this feature will greatly facilitate and relief the development process for almost each developer using Node.js and NPM.

**Comment by ljharb** [comment]
That’s what automatically happens if peerDependencies are defined correctly by every package that needs them. If they’re not, how would npm know what’s compatible?

**Reply by TetianaNazarova** [reply]
I will test it and answer you. 

But as I understand peerDependencies just give you an error, but **what I'm struggling for that npm will throw an solution with the compatible versions** of the installed package and another packages in package.json.

**Reply by TetianaNazarova** [reply]
If NPM doesn't know the compatible version of packages, that will not throw the error of incompatibility. Then, who knows it? Author of the installed package? But if he/she could know with what packages his/her package will be used in advance? 

I guess if npm throws an error like 'Package A needs the version of package B to be > 10 and < 13.2.1' it can know the exact compatible version for these packages and it can give such information to the developer (I'm not saying about auto changing versions of packages in packages.json).

**Reply by ljharb** [reply]
Yes, the author of the package is the only person who knows compatibility, and it’s their responsibility to declare it with peerDependencies and engines. npm already respects these declarations.
